src/coreclr/src/tools/Common/Sorting/MergeSortCore.cs

```diff
@@ -41,17 +41,15 @@ private static async Task ParallelSort(TDataStructure arrayToSort, int index, in
                 TDataStructureAccessor accessor = default(TDataStructureAccessor);
                 int halfLen = length / 2;
 
-                TaskCompletionSource<bool> rightSortComplete = new System.Threading.Tasks.TaskCompletionSource<bool>();
-                _ = Task.Run(async () =>
+                Task rightSortTask = Task.Run(async () =>
                 {
                     await ParallelSort(arrayToSort, index + halfLen, length - halfLen, comparer);
-                    rightSortComplete.SetResult(true);
```

Comment:
There was nothing to signal the TaskCompletionSource when the child task threw exception. It resulted into hang. Awaiting the child task directly fixes the error propagation.

Was there a reason why this used TaskCompletionSource instead of awaiting the child taks directly that I am missing?

Reply:
+@davidwrighton since he added support for ParallelSort